built_value_test/lib/matcher.dart

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) 2017, Google Inc. Please see the AUTHORS file for details.
+// All rights reserved. Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/built_value.dart';
+import 'package:test/test.dart';
+
+/// Returns a matcher that matches if the value is structurally equal to
+/// [expected].
+///
+/// Improves on a simple equality test by offering a detailed mismatch message.
+Matcher equalsBuilt(Built expected) => new _BuiltValueMatcher(expected);
+
+/// Matcher for [Built] instances.
+///
+/// Converts instances to maps using [_CapturingToStringHelper] then compares
+/// using [equals], which does deep comparison on maps.
+class _BuiltValueMatcher implements Matcher {
+  final Built _expected;
+  final Matcher _delegate;
+
+  _BuiltValueMatcher(this._expected) : _delegate = equals(_toMap(_expected));
+
+  @override
+  Description describe(Description description) =>
+      _delegate.describe(description);
+
+  @override
+  Description describeMismatch(dynamic item, Description mismatchDescription,
+      Map matchState, bool verbose) {
+    if (_expected.runtimeType != item.runtimeType)
+      return mismatchDescription.add('is the wrong type');
+
+    return _delegate.describeMismatch(
+        _toMap(item), mismatchDescription, matchState, verbose);
+  }
+
+  @override
+  bool matches(dynamic item, Map matchState) {
+    if (_expected.runtimeType != item.runtimeType) return false;
+
+    return _delegate.matches(_toMap(item), matchState);
+  }
+}
+
+/// Converts a Built to a map.
+Map _toMap(Object built) {
```

Comment:
It's <String, Object>. Done.

(If it was <String, String> we'd end up checking for string equality instead of actual equality.)